add a piecewise quaternion trajectory based on slerp.

siyuanfeng-tri opened the pull request:
1. add a piecewise quaternion trajectory based on slerp.
2. also changed the getSegmentIndex to binary search on time instead of
linear serach.

@hongkai-dai for feature review, @tkoolen for opinions. thanks guys.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4432)
<!-- Reviewable:end -->

jamiesnape commented:
@drake-jenkins-bot retest this please

tkoolen commented:
Instead of using time directly as the interpolation parameter, you may want to use a sigmoid function of time that starts and ends with zero derivative (and possibly second derivative). With this implementation, angular velocity changes discontinuously at every knot. Either that or use a squad type approach if you want nonzero angular velocities at the knots.

siyuanfeng-tri commented:
yeah.. i was thinking about if i want to squad or something like that.
but i really just want a thing as simple as slerp. also for the stuff we do, the change in orientation is minimal anyway. 
i will most likely add a more continuous thing later on. 

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYVrTNLZp7rSEoepp1Y:b-glmpqg)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Check point

---

Reviewed 3 of 9 files at r1.
Review status: 3 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/trajectories/piecewise_function.cc, line 69 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYVvPhRn_koYRzwGaGx:-KYVvPhRn_koYRzwGaGy:b-tho4s5) ([raw file](https://github.com/robotlocomotion/drake/blob/8327a4816c11a61b2a11e7a95d32878aa19daa42/drake/common/trajectories/piecewise_function.cc#L69)):*
> ```cpp
>   else
>     return mid;
> }
> ```

This will return `segment_times[index-1]` if `time = segment_times[index] - epsilon`. I am not sure the context to use this function, would it help if you add an epsilon here, like
```C++
if (time < segment_times[mid] - epsilon)
  return GetSegmentIndexRecursive(time, start, mid);
else if (time > segment_times[mid] + epsilon)
   return GetSegmentIndexRecursive(time, mid, end);
else
   return mid;
```

---

*[drake/math/quaternion.h, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYVtUo3QFzDcIrqp-J_:-KYVtUo3QFzDcIrqp-Ja:b75c9m5) ([raw file](https://github.com/robotlocomotion/drake/blob/8327a4816c11a61b2a11e7a95d32878aa19daa42/drake/math/quaternion.h#L20)):*
> ```objc
> /**
>  * @Returns a quaternion that represents the same orientation as `q1`,
>  * and is the "closest" to `q0`.
> ```

Maybe say it is "closest" in the geodesic distance on the unit sphere.

---

*[drake/math/quaternion.h, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYVuHceZoiekcXIku07:-KYVuHceZoiekcXIku08:b-ufcc8f) ([raw file](https://github.com/robotlocomotion/drake/blob/8327a4816c11a61b2a11e7a95d32878aa19daa42/drake/math/quaternion.h#L23)):*
> ```objc
>  */
> template <typename Scalar>
> const Eigen::Quaternion<Scalar> ClosetQuaternion(
> ```

Why returning the result by const? It will forbid using rvalue reference. http://stackoverflow.com/questions/8716330/purpose-of-returning-by-const-value

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYVw4XZge0YrHGZ2r3m:b-3lrsgd)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 3 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/trajectories/piecewise_function.cc, line 69 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYVvPhRn_koYRzwGaGx:-KYYqoXFcylKTfjRgnq3:b-fdllso) ([raw file](https://github.com/robotlocomotion/drake/blob/8327a4816c11a61b2a11e7a95d32878aa19daa42/drake/common/trajectories/piecewise_function.cc#L69)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

This will return `segment_times[index-1]` if `time = segment_times[index] - epsilon`. I am not sure the context to use this function, would it help if you add an epsilon here, like
```C++
if (time < segment_times[mid] - epsilon)
  return GetSegmentIndexRecursive(time, start, mid);
else if (time > segment_times[mid] + epsilon)
   return GetSegmentIndexRecursive(time, mid, end);
else
   return mid;
```
</blockquote></details>

my intention is to return index-1 if that's the case. there is a test case for this as well. i want to keep the same behavior as the previous while loop. 

---

*[drake/math/quaternion.h, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYVtUo3QFzDcIrqp-J_:-KYYq9IMPSPApCAKFzsZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8327a4816c11a61b2a11e7a95d32878aa19daa42/drake/math/quaternion.h#L20)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Maybe say it is "closest" in the geodesic distance on the unit sphere.
</blockquote></details>

Done.

---

*[drake/math/quaternion.h, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYVuHceZoiekcXIku07:-KYYqACV5v6whxpySrnQ:b-3fihxf) ([raw file](https://github.com/robotlocomotion/drake/blob/8327a4816c11a61b2a11e7a95d32878aa19daa42/drake/math/quaternion.h#L23)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Why returning the result by const? It will forbid using rvalue reference. http://stackoverflow.com/questions/8716330/purpose-of-returning-by-const-value
</blockquote></details>

opps, that was not intended. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@hongkai-dai made the api change, can you take another look when you have a chance. thanks

---

Review status: 2 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KY_j1J43FOXh9y9a-DS:bp3z6e)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Thanks for the change! The code is beautiful.

---

Reviewed 4 of 9 files at r1, 1 of 1 files at r3, 2 of 3 files at r4.
Review status: 8 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 43 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KY_kbrFeRHZCpgqNH0r:-KY_kbrFeRHZCpgqNH0s:b-wyor03) ([raw file](https://github.com/robotlocomotion/drake/blob/53618c3444d82f880bd492d26f6180873ba339b8/drake/common/trajectories/piecewise_quaternion.h#L43)):*
> ```objc
>    * or breaks have length < 2.
>    */
>   PiecewiseQuaternionSlerp(
> ```

This should not have a enforce_closest option. Only the quaternion and angle-axis have the ambiguity that there are two representations for the same orientation. 

Is it OK if we offer slerp function for all three rotation representations: quaternion, angle-axis and rotation, with class
`PiecewiseQuaternionSlerp`, `PiecewiseAngleAxisSlerp` and `PiecewiseRotationMatrixSlerp`, they may all be derived from a base class `PiecewiseOrientationSlerp()`.

The first two class can need to consider the closest path problem, the rotation matrix representation does not, each of these class have a method
`get_orientation_knots()` instead of the current `get_quaternion_knots()`.
```
eigen_aligned_std_vector<AngleAxis<Scalar>>&  PiecewiseAngleAxisSlerp::get_orientation_knots();
eigen_aligned_std_vector<Quaternion<Scalar>>& PiecewiseQuaternionSlerp::get_orientation_knots();
eigen_aligned_std_vector<Matrix3d<Scalar>>&   PiecewiseRotationMatrixSlerp::get_orientation_knots();
```

---

*[drake/common/trajectories/test/piecewise_function_test.cc, line 63 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KY_paY54Z5ka0WM4sC2:-KY_paY6Ju9kzaHb45sh:b-vq1w43) ([raw file](https://github.com/robotlocomotion/drake/blob/53618c3444d82f880bd492d26f6180873ba339b8/drake/common/trajectories/test/piecewise_function_test.cc#L63)):*
> ```cpp
>     int idx = traj.getSegmentIndex(t);
>     EXPECT_GE(idx, 0);
>     EXPECT_TRUE(idx < traj.getNumberOfSegments());
> ```

`EXPECT_LT`

---

*[drake/common/trajectories/test/piecewise_quaternion_test.cc, line 160 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KY_tdehBgHsq-d-y9zt:-KY_tdehBgHsq-d-y9zu:b-pmyq2p) ([raw file](https://github.com/robotlocomotion/drake/blob/53618c3444d82f880bd492d26f6180873ba339b8/drake/common/trajectories/test/piecewise_quaternion_test.cc#L160)):*
> ```cpp
>     // Angular velocity goes the "other" direction if we want the "longer"
>     // distance between the two quaterions.
>     double omega = angleDiff(ang[i], ang[i + 1]);
> ```

I prefer not to call it `omega`, but probably `rotation_angle`, `omega` means angular velocity to me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KY_trsbESOy1cXXcY2J:by4p706)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
I think we should add one more test case for doing slerp interpolation between two identical quaternion to cover this corner case.

---

Review status: 8 of 9 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KY_z5jLKUsVp52CYTM5:b-n9tklr)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 5 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 43 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KY_kbrFeRHZCpgqNH0r:-KYa_EWd2aAJaUuMapTO:bjldbep) ([raw file](https://github.com/robotlocomotion/drake/blob/53618c3444d82f880bd492d26f6180873ba339b8/drake/common/trajectories/piecewise_quaternion.h#L43)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

This should not have a enforce_closest option. Only the quaternion and angle-axis have the ambiguity that there are two representations for the same orientation. 

Is it OK if we offer slerp function for all three rotation representations: quaternion, angle-axis and rotation, with class
`PiecewiseQuaternionSlerp`, `PiecewiseAngleAxisSlerp` and `PiecewiseRotationMatrixSlerp`, they may all be derived from a base class `PiecewiseOrientationSlerp()`.

The first two class can need to consider the closest path problem, the rotation matrix representation does not, each of these class have a method
`get_orientation_knots()` instead of the current `get_quaternion_knots()`.
```
eigen_aligned_std_vector<AngleAxis<Scalar>>&  PiecewiseAngleAxisSlerp::get_orientation_knots();
eigen_aligned_std_vector<Quaternion<Scalar>>& PiecewiseQuaternionSlerp::get_orientation_knots();
eigen_aligned_std_vector<Matrix3d<Scalar>>&   PiecewiseRotationMatrixSlerp::get_orientation_knots();
```
</blockquote></details>

good point for the matrix rep!  
i am going to make it enforce shortest for the matrix case. 
let's not go super excited about this yet.. i don't really see a demand for make that clear an distinction, and if there is actual need. we can think more about it. i want to have a simple, easy to understand rotation interpolation that i can trust for now :) 

---

*[drake/common/trajectories/test/piecewise_function_test.cc, line 63 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KY_paY54Z5ka0WM4sC2:-KYaaJrgjCXyt9uov9xq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/53618c3444d82f880bd492d26f6180873ba339b8/drake/common/trajectories/test/piecewise_function_test.cc#L63)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`EXPECT_LT`
</blockquote></details>

Done.

---

*[drake/common/trajectories/test/piecewise_quaternion_test.cc, line 160 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KY_tdehBgHsq-d-y9zt:-KYaaN2oeWPxPZHH_vUM:b2jj9vy) ([raw file](https://github.com/robotlocomotion/drake/blob/53618c3444d82f880bd492d26f6180873ba339b8/drake/common/trajectories/test/piecewise_quaternion_test.cc#L160)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I prefer not to call it `omega`, but probably `rotation_angle`, `omega` means angular velocity to me.
</blockquote></details>

it's eventually divided by time, so i think omega is fine. reviewable chopped that part off. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
How does the slerp behave if we give two quaternions with identical orientation, but the sign is flipped? Namely `slerp(q, -q, t)`, and choose not to take the closest path? In that case there are infinitely many rotation axis. 

---

Review status: 3 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYjAwVhgSwfUhQHCg_4:bhk41to)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
huge oooops!
i was looking at the eigen code for this, /home/sfeng/code/drake/build/install/include/eigen3/Eigen/src/Geometry/Quaternion.h
and it turns out, eigen is enforcing the closest anyway. there is a dot product check in the source code...
i also did a quick check to just make sure that is the case, and q0.slerp(t, q1) = q0.slerp(t, -q1). 

i didn't check the (omega * dt) * q0 == q1 test for the non-shortest distance case, so i didn't caught this earlier... 

---

Review status: 3 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYjG8QXZebBZ5_t9bzM:b4cel8r)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
i am going to change it back to assuming the shortest distance version. i find it easier to work with, if we really want to force it to go the other way, we can always manually insert another knot halfway when designing the knots. 

---

Review status: 3 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYjLIUIj6eEc6E7EPmN:bc7yi8f)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Wow, that is interesting.

Reading Eigen's source code http://eigen.tuxfamily.org/dox/Quaternion_8h_source.html, I am not very happy with Eigen's implementation of slerp. It calls 
```
theta = acos(absD)
```
As far as I know, `acos(x)` function is numerically bad near x=1, where the gradient is almost infinity. Eigen has a special case when `x > 1 - epesilon`. But in my experience `acos(x)` is bad even when `x = 1 - 1E-8`.  I will talk with @mitiguy to see if we can find a better solution to this. At the moment, let's just use Eigen's implementation.

---

Review status: 3 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYjN5UkCJJT_S_xl5ny:b-t622jh)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
sounds good to me. thanks for being so thorough :) 

---

Review status: 3 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYjPZN389bN7G6gaNd8:b-sjyq8p)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 3 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYjnjfzGY66RqU6RZB2:bnfp4nl)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@sherm1 for platform review. thanks!

sherm1 commented:
Platform pass done. Looks very nice, some comments below - PTAL. Please generate the doxygen documentation and take a look at it to make sure it is formatting as you want it to. And be sure to start every doxygen block with a short sentence that doxygen will use as the "brief" for the class or method (prior to any doxygen commands like `@param` or `@return`).

While you are making substantial changes to this class, please consider fixing the non-compliant method names like `getStartTime()` to be style guide compliant like your own code is. It would be nice to get that over with while you are in this code.

---

Reviewed 2 of 9 files at r1, 1 of 1 files at r3, 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/common/trajectories/CMakeLists.txt, line 1 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkUNxJ1C3tfo5O7e8V:-KYkUNxJ1C3tfo5O7e8W:b-ajbb2u) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/CMakeLists.txt#L1)):*
> ```CMake
> add_definitions(-g)
> ```

What is this doing here? Looks like it is forcing compilation in Debug mode?

---

*[drake/common/trajectories/piecewise_function.h, line 56 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkVJgZEBIsTQ8GF5KH:-KYkVJgZEBIsTQ8GF5KI:b5j51ba) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_function.h#L56)):*
> ```objc
> 
>  private:
>   size_t GetSegmentIndexRecursive(double time, size_t start, size_t end) const;
> ```

The [style guide](https://google.github.io/styleguide/cppguide.html#Integer_Types) prohibits use of unsigned integers in most cases. Also see the Drake [clarification](http://drake.mit.edu/code_style_guide.html#id15) about `size_t` specifically. It is rarely allowed and this doesn't appear to be one of the acceptable exceptions. If you need a large value you can use `int64_t`, otherwise just use `int`.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 15 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkWNmPK90_0hf3CmCl:-KYkWNmPK90_0hf3CmCm:b-tah3v0) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L15)):*
> ```objc
>  * using piecewise slerp (spherical linear interpolation).
>  * All the quaternions, angular velocity, angular acceleration are assumed to
>  * be in the same reference frame.
> ```

A quaternion representing a rotation R_AB giving B's orientation in A contains a vector that is numerically identically in both frames so could be considered to be expressed in either A or B. Angular velocity and acceleration, OTOH, must be given in *either* frame A or B but will be numerically different depending on which one. So this sentence is ambiguous. If it matters in which frame these are expressed, which one do you require here?

---

*[drake/common/trajectories/piecewise_quaternion.h, line 22 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkXLSRmP8qF050jq5T:-KYkXLSSnERhvG5NYS5j:b-4auqvo) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L22)):*
> ```objc
>  * have the shortest geodesic distance on the unit sphere.
>  */
> template <typename Scalar = double>
> ```

Please document the template argument (can use `@tparam`). Is this the usual Eigen scalar type like double or AutoDiffScalar?

---

*[drake/common/trajectories/piecewise_quaternion.h, line 52 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkXi4X8tp-0hIDevT-:-KYkXi4Y7x4Nd6hOZ6Rr:b-iidmmq) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L52)):*
> ```objc
>   PiecewiseQuaternionSlerp(
>       const std::vector<double>& breaks,
>       const eigen_aligned_std_vector<AngleAxis<Scalar>>& ang_axises);
> ```

BTW, the plural of `axis` is `axes`.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 58 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkY1YR8OtR9aE50OcY:-KYkY1YR8OtR9aE50OcZ:bt1a8mm) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L58)):*
> ```objc
> 
>   /**
>    * @param t Time for interpolation.
> ```

All doxygen comments should begin with a short sentence that doxygen can use as the "brief" description (that is, before the first `@param`). Here and below.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 78 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkYDXbNkymo-4fiJyh:-KYkYDXbNkymo-4fiJyi:b-gnxkrn) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L78)):*
> ```objc
> 
>   /**
>    * Note: the returned quaternions might be different from the ones that are
> ```

Should start with a short description before the note. Also you can use doxygen's `@note` command if you want the note to stand out.

---

*[drake/common/trajectories/test/piecewise_function_test.cc, line 8 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkaPL-n9Te07yCyDW9:-KYkaPL-n9Te07yCyDWA:bb7phvt) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/test/piecewise_function_test.cc#L8)):*
> ```cpp
> #include "gtest/gtest.h"
> 
> // Dummy implmentation of PiecewiseFunction to test the basic indexing
> ```

typo

---

*[drake/math/quaternion.h, line 19 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYk_FFN8nDZDpO0P6oB:-KYk_FFN8nDZDpO0P6oC:bgyumuq) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/math/quaternion.h#L19)):*
> ```objc
> 
> /**
>  * @Returns a quaternion that represents the same orientation as `q1`,
> ```

Needs a short sentence to be used as doxygen "brief" (before the first `@` command). The sentence you have could work for the brief and you could then leave out the `@returns` command altogether. (Also I'm not sure if the capitalized version is recognized.)

---

*[drake/math/quaternion.h, line 22 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkZJId9qdFlqWcK9WQ:-KYkZJId9qdFlqWcK9WR:b-9o3nno) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/math/quaternion.h#L22)):*
> ```objc
>  * and has the "shortest" geodesic distance on the unit sphere to `q0`.
>  */
> template <typename Scalar> Eigen::Quaternion<Scalar> ClosetQuaternion(
> ```

Typo Closet -> Closest (unless you mean it is hiding the fact that it's a quaternion :).

---

*[drake/math/quaternion.h, line 30 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkZXnd2nfUMuLXTcsp:-KYkZXnd2nfUMuLXTcsq:b-9b721e) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/math/quaternion.h#L30)):*
> ```objc
>     q.x() *= -1;
>     q.y() *= -1;
>     q.z() *= -1;
> ```

BTW, does `q *= -1` or `q = -q` work? I think those would be clearer.

---

*[drake/math/quaternion.h, line 32 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkZnXogL3HUQtgetfd:-KYkZnXpIyXrhTSdlQ57:b-maelzk) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/math/quaternion.h#L32)):*
> ```objc
>     q.z() *= -1;
>   }
>   q.normalize();
> ```

Is this necessary? If so, it should be mentioned in the documentation that the method normalizes the result. But it's expensive and an alternative would be just to say the returned value q has the same norm as the input q1. Either way is fine depending on what you want the method to do, but the documentation should be clear about it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYk_eolKX3dqEiskWhe:b-jlzz06)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
i will open another pr to change the names. it's probably going to touch a lot of files. 

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/common/trajectories/CMakeLists.txt, line 1 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkUNxJ1C3tfo5O7e8V:-KYkoZ5sJDiJACb5oQkh:b-e6uprj) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/CMakeLists.txt#L1)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

What is this doing here? Looks like it is forcing compilation in Debug mode?
</blockquote></details>

some leftover dirty laundry  

---

*[drake/common/trajectories/piecewise_function.h, line 56 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkVJgZEBIsTQ8GF5KH:-KYkpfZG9D-Qm4APsDOX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_function.h#L56)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

The [style guide](https://google.github.io/styleguide/cppguide.html#Integer_Types) prohibits use of unsigned integers in most cases. Also see the Drake [clarification](http://drake.mit.edu/code_style_guide.html#id15) about `size_t` specifically. It is rarely allowed and this doesn't appear to be one of the acceptable exceptions. If you need a large value you can use `int64_t`, otherwise just use `int`.
</blockquote></details>

Done.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 15 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkWNmPK90_0hf3CmCl:-KYkthUKmLZgDmGgianr:b-q0y3ar) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L15)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

A quaternion representing a rotation R_AB giving B's orientation in A contains a vector that is numerically identically in both frames so could be considered to be expressed in either A or B. Angular velocity and acceleration, OTOH, must be given in *either* frame A or B but will be numerically different depending on which one. So this sentence is ambiguous. If it matters in which frame these are expressed, which one do you require here?
</blockquote></details>

in most of my use cases, the implicit base frame is the world frame (or whatever common frame), and the velocity and acceleration are all expressed in that frame. 
so everything this trajectory holds is expressed in that common frame. i didn't realized there is any ambiguity here. 
how do i make this clearer? 

---

*[drake/common/trajectories/piecewise_quaternion.h, line 22 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkXLSRmP8qF050jq5T:-KYky_1lROWiczY3d9G8:b-bjd0s8) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L22)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Please document the template argument (can use `@tparam`). Is this the usual Eigen scalar type like double or AutoDiffScalar?
</blockquote></details>

double / float. i honestly don't know if it works for autodiff. 

---

*[drake/common/trajectories/piecewise_quaternion.h, line 52 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkXi4X8tp-0hIDevT-:-KYkq6sRiMSy0RMEDqVg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L52)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, the plural of `axis` is `axes`.
</blockquote></details>

Done.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 58 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkY1YR8OtR9aE50OcY:-KYkqZ63t6eXWBK5fDLj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L58)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

All doxygen comments should begin with a short sentence that doxygen can use as the "brief" description (that is, before the first `@param`). Here and below.
</blockquote></details>

Done.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 78 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkYDXbNkymo-4fiJyh:-KYkqwu8mdHGKHNwcSIb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L78)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Should start with a short description before the note. Also you can use doxygen's `@note` command if you want the note to stand out.
</blockquote></details>

Done.

---

*[drake/common/trajectories/test/piecewise_function_test.cc, line 8 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkaPL-n9Te07yCyDW9:-KYkr3svk3NEt5wR0ezN:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/test/piecewise_function_test.cc#L8)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

typo
</blockquote></details>

Done.

---

*[drake/math/quaternion.h, line 19 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYk_FFN8nDZDpO0P6oB:-KYkstbB6su6o1vh4Rot:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/math/quaternion.h#L19)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Needs a short sentence to be used as doxygen "brief" (before the first `@` command). The sentence you have could work for the brief and you could then leave out the `@returns` command altogether. (Also I'm not sure if the capitalized version is recognized.)
</blockquote></details>

Done.

---

*[drake/math/quaternion.h, line 22 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkZJId9qdFlqWcK9WQ:-KYkstLv8RBKCfYIWqur:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/math/quaternion.h#L22)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Typo Closet -> Closest (unless you mean it is hiding the fact that it's a quaternion :).
</blockquote></details>

Done.

---

*[drake/math/quaternion.h, line 30 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkZXnd2nfUMuLXTcsp:-KYksv0xXIHOdOrUlCTs:b33bxnh) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/math/quaternion.h#L30)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, does `q *= -1` or `q = -q` work? I think those would be clearer.
</blockquote></details>

changed to q.coeffs. 

---

*[drake/math/quaternion.h, line 32 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkZnXogL3HUQtgetfd:-KYkszK-qqQ972ylzQ3S:bsm1l98) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/math/quaternion.h#L32)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Is this necessary? If so, it should be mentioned in the documentation that the method normalizes the result. But it's expensive and an alternative would be just to say the returned value q has the same norm as the input q1. Either way is fine depending on what you want the method to do, but the documentation should be clear about it.
</blockquote></details>

i added a part in the doc saying it's normalized. i liked it normalized just to have a peace of mind. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYktQ2DfTKhOvAwpqIW:bz3tg7r)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
A few more comments.

---

Reviewed 5 of 7 files at r6.
Review status: 7 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 15 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkWNmPK90_0hf3CmCl:-KYns07fmLUXNXQQa0TD:brex9z3) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L15)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

in most of my use cases, the implicit base frame is the world frame (or whatever common frame), and the velocity and acceleration are all expressed in that frame. 
so everything this trajectory holds is expressed in that common frame. i didn't realized there is any ambiguity here. 
how do i make this clearer? 
</blockquote></details>

Assuming the following is your intention, you could say something like:
Given a quaternion representing rotation matrix R_PC, that is, a "child" frame C's orientation in a "parent" frame P (often World), this class expects the angular velocity vector to be C's angular velocity in P, expressed in the P frame, and expects angular acceleration to follow the same convention.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 22 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkXLSRmP8qF050jq5T:-KYntVgli_stwMsTBwal:b-iphp3j) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L22)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

double / float. i honestly don't know if it works for autodiff. 
</blockquote></details>

The rest of Drake only works in `double` so this template parameter would only be useful if it can by used with the fancier scalar types. If this will ultimately need to be differentiated, please try instantiating with an AutoDiffScalar to see if you can easily get it to work. If not, document the currently allowable template arguments and add a TODO.

---

*[drake/math/quaternion.h, line 19 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYo0bUZbX-Pr1wp8xGu:-KYo0bUZbX-Pr1wp8xGv:b-tf59zr) ([raw file](https://github.com/robotlocomotion/drake/blob/3e24748197b26879980e9dc38f704ae2785ff74f/drake/math/quaternion.h#L19)):*
> ```objc
> 
> /**
>  * Returns an unit quaternion that represents the same orientation as `q1`,
> ```

BTW, "an unit" -> "a unit".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYo1xUK2dAU62QPlxZd:byu0gl7)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 2 of 7 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 7 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 15 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkWNmPK90_0hf3CmCl:-KYoHmvcsUTPB48e4WMe:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L15)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Assuming the following is your intention, you could say something like:
Given a quaternion representing rotation matrix R_PC, that is, a "child" frame C's orientation in a "parent" frame P (often World), this class expects the angular velocity vector to be C's angular velocity in P, expressed in the P frame, and expects angular acceleration to follow the same convention.
</blockquote></details>

Done.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 22 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYkXLSRmP8qF050jq5T:-KYoIB678DeWBwaYhbw2:b-knv1ek) ([raw file](https://github.com/robotlocomotion/drake/blob/ecfc249489178c876005978d5dc5f8d32fdb7d19/drake/common/trajectories/piecewise_quaternion.h#L22)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

The rest of Drake only works in `double` so this template parameter would only be useful if it can by used with the fancier scalar types. If this will ultimately need to be differentiated, please try instantiating with an AutoDiffScalar to see if you can easily get it to work. If not, document the currently allowable template arguments and add a TODO.
</blockquote></details>

added a todo.

---

*[drake/math/quaternion.h, line 19 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYo0bUZbX-Pr1wp8xGu:-KYoIIVETX8J3p53HJhy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3e24748197b26879980e9dc38f704ae2785ff74f/drake/math/quaternion.h#L19)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, "an unit" -> "a unit".
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
A couple of comments, BTWed for non-blockingness but please address. Then: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 15 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYoxLCR6jO1yvcucSBk:-KYoxLCR6jO1yvcucSBl:bwkhihv) ([raw file](https://github.com/robotlocomotion/drake/blob/1247a440c230fe99826fc6e09857e02783b5f33d/drake/common/trajectories/piecewise_quaternion.h#L15)):*
> ```objc
>  * using piecewise slerp (spherical linear interpolation).
>  * All the orientation knots are expected to be with respect to the same
>  * parent reference frame, i.e. q_i represents the rotation R^p_i,
> ```

BTW, we are attempted to standardize on ascii notation for rotation matrices where
```
A B
 R
```
is written `R_AB` because it allows relatively mindless letter-matching to make sure rotation expressions are correct (like `v_A = R_AB*v_B`) where v_A means "vector v expressed in frame A" etc. So here I would write something like: "q_i represents the rotation `R_PBi`for the orientation of frame B at knot i, in a fixed parent frame P. (Capital letters are conventional for frames.)

---

*[drake/common/trajectories/piecewise_quaternion.h, line 28 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYp-qYOQbuWLZ10lO2j:-KYp-qYOQbuWLZ10lO2k:bcht7vk) ([raw file](https://github.com/robotlocomotion/drake/blob/1247a440c230fe99826fc6e09857e02783b5f33d/drake/common/trajectories/piecewise_quaternion.h#L28)):*
> ```objc
>  * @tparam Scalar, double.
>  *
>  * TODO(siyuan.feng): check if this works for AutoDiffScalar.
> ```

BTW, TODOs should not be in doxygen comments -- just use a developer comment like `// TODO() ....`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYp05Y3Xj-2kpgxYSty:b-atftgc)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 15 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYoxLCR6jO1yvcucSBk:-KYq93wTNv9mDw8vKFso:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1247a440c230fe99826fc6e09857e02783b5f33d/drake/common/trajectories/piecewise_quaternion.h#L15)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, we are attempted to standardize on ascii notation for rotation matrices where
```
A B
 R
```
is written `R_AB` because it allows relatively mindless letter-matching to make sure rotation expressions are correct (like `v_A = R_AB*v_B`) where v_A means "vector v expressed in frame A" etc. So here I would write something like: "q_i represents the rotation `R_PBi`for the orientation of frame B at knot i, in a fixed parent frame P. (Capital letters are conventional for frames.)
</blockquote></details>

Done.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 28 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-KYp-qYOQbuWLZ10lO2j:-KYq97MTwrQe20OU8iRj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1247a440c230fe99826fc6e09857e02783b5f33d/drake/common/trajectories/piecewise_quaternion.h#L28)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, TODOs should not be in doxygen comments -- just use a developer comment like `// TODO() ....`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@sherm1 sorry for the late reply, was chasing down some rabbit hole in the robot land. i see two unresolved issues regarding the comments. thanks!

---

Review status: 8 of 9 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYq9d7m3RbWvStFaJte:b-ib05uu)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Thanks, Siyuan -- <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Review status: 8 of 9 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4432#-:-KYqYRfO16DrGN3Wa3CJ:baiqg8x)*
<!-- Sent from Reviewable.io -->
